### PR TITLE
Feature example steps

### PR DESCRIPTION
**Pull request by holke**
Wrote tutorials that lead a user or future developer step-by-step.

This PR is to include the first tutorials into develop.
If you think there are tutorials missing, we should merge nevertheless and continue implementing the new tutorials on this branch.

Please also review the WIKI pages that go along with the examples.

https://github.com/holke/t8code/wiki/Step-0---Hello-World
https://github.com/holke/t8code/wiki/Step-1---Creating-a-coarse-mesh
https://github.com/holke/t8code/wiki/Step-2---Creating-a-uniform-forest
https://github.com/holke/t8code/wiki/Step-3---Adapting-a-forest
https://github.com/holke/t8code/wiki/Step-4---Partition,-Balance,-Ghost

Note: #74 depends on this PR

### CODE REVIEW

**Comment by holke** [comment]
Edit: This was a bit too much an contained different topics, so i split it up.

The search example required a point_inside functionality to be implemented first.
This function should have a code review on its own, so it moved it to #73 and removed the search example from this pr.
The search example now has its own pr #74

**Comment by Davknapp** [comment]
In the example t8_step1_coarsemesh.c, when changing line 60 to

  cmesh = t8_cmesh_new_hypercube (T8_ECLASS_TET, comm, 0, 1, 0);
  
  and running with mpirun -np 2 ./t8_step1_coarsemesh
  it raises the Assertion at src/t8_cmesh/t8_cmesh_vtk.c:160 "vertices != NULL!

**Comment by Davknapp** [comment]
In the example t8_step1_coarsemesh.c

The console always prints "Created cmesh with 6 tetrahedra", even if the tree-shape has been changed in line 60.
Would it be to confusing for the user to print a statement depending on the tree-shape? If no, we should change that.

**Comment by holke** [comment]
> 
> 
> In the example t8_step1_coarsemesh.c
> 
> The console always prints "Created cmesh with 6 tetrahedra", even if the tree-shape has been changed in line 60.
> Would it be to confusing for the user to print a statement depending on the tree-shape? If no, we should change that.

Yes, we should be able to change that.
A message containing the number of trees is easy, the tree type depends on the choosen cmesh...will look into this.

**Comment by Davknapp** [comment]
I also think, that it would be interesting to print the number of elements of each process in step2 (line 163). Right now, it is a t8_global_productionf and should be changed to t8_productionf in my opinion.

**Comment by holke** [comment]
> In the example t8_step1_coarsemesh.c, when changing line 60 to
> 
> cmesh = t8_cmesh_new_hypercube (T8_ECLASS_TET, comm, 0, 1, 0);
> 
> and running with mpirun -np 2 ./t8_step1_coarsemesh
> it raises the Assertion at src/t8_cmesh/t8_cmesh_vtk.c:160 "vertices != NULL!

This usually happens when the cmesh does not have geometry information.
Thats weird because the hypercube cmesh should have this info...

**Comment by holke** [comment]
> I also think, that it would be interesting to print the number of elements of each process in step2 (line 163). Right now, it is a t8_global_productionf and should be changed to t8_productionf in my opinion.
> Some output like:
> [step 2] local number of elements on proc 0: xy
> [step 2] local number of elements on proc 1: xy
> would be the best

good idea

File: example/tutorials/README.md
Comment (Davknapp) [inline example/tutorials/README.md:22]: Is step6 in production?

File: example/tutorials/t8_step1_coarsemesh.c
Comment (Davknapp) [inline example/tutorials/t8_step1_coarsemesh.c:60]: raises Assertion with  cmesh = t8_cmesh_new_hypercube (T8_ECLASS_TET, comm, 0, 1, 0);

File: example/tutorials/t8_step2_uniform_forest.cxx
Comment (Davknapp) [inline example/tutorials/t8_step2_uniform_forest.cxx:163]:
```suggestion
  t8_productionf (" [step2] Local number of elements:\t\t%i\n",
```

File: example/tutorials/t8_step3.h
Comment (Davknapp) [inline example/tutorials/t8_step3.h:28]: Most information about this example is in t8_step3_adapt_forest.cxx

```suggestion
 * See \ref t8_step3_adapt_forest.cxx for more details.
```

**Review by Davknapp — CHANGES_REQUESTED** [review]
All in all very good and clean Code. Some files have a copy-paste error in the comments, which should be changed. 
The Assertion-error should be solved before merging

**Comment by holke** [comment]
> In the example t8_step1_coarsemesh.c, when changing line 60 to
> 
> cmesh = t8_cmesh_new_hypercube (T8_ECLASS_TET, comm, 0, 1, 0);
> 
> and running with mpirun -np 2 ./t8_step1_coarsemesh
> it raises the Assertion at src/t8_cmesh/t8_cmesh_vtk.c:160 "vertices != NULL!

Damn, this seems to be a major bug. See #79.
Would shall we do? Wait until the bug is resolved?
Or merge this regardless, since the bug is in develop nevertheless?

**Comment by holke** [comment]
Okay for now i added a warning that is printed when the hypercube cmesh is created.
Do you believe this is a viable solution until the cmesh bug is resolved?

File: example/tutorials/t8_step1_coarsemesh.c
Comment (holke) [reply example/tutorials/t8_step1_coarsemesh.c:60]: Damn, this seems to be a major bug. See #79 and #80.
Would shall we do? Wait until the bug is resolved?
Or merge this regardless, since the bug is in develop nevertheless?

Edit: Okay for now i added a warning that is printed when the hypercube cmesh is created.
Do you believe this is a viable solution until the cmesh bug is resolved?

**Comment by holke** [comment]
Note to future self: Add link to the wiki pages to each example.

File: example/tutorials/t8_step1_coarsemesh.c
Comment (Davknapp) [reply example/tutorials/t8_step1_coarsemesh.c:60]: Since you opened issues regarding this bug, I think it is a viable solution.

**Review by Davknapp — APPROVED** [review]

**Review by Davknapp — COMMENTED** [review]
I think we can merge this, after attacking my comments regarding some output of the tutorials

File: example/tutorials/t8_step2_uniform_forest.cxx
Comment (holke) [reply example/tutorials/t8_step2_uniform_forest.cxx:163]: See the note in the Wiki on this: "Note that due to using t8_global_productionf opposed to t8_productionf only rank 0 will print its local number of elements."
I will leave it as is.

**Comment by holke** [comment]
> I also think, that it would be interesting to print the number of elements of each process in step2 (line 163). Right now, it is a t8_global_productionf and should be changed to t8_productionf in my opinion.
> Some output like:
> [step 2] local number of elements on proc 0: xy
> [step 2] local number of elements on proc 1: xy
> would be the best

This was deliberate to 
a) Not overwhelm the user with output on their first try
b) Provide an extra challenge to the user to implement it themselves.

There is a section in the Wiki adressing this.

**Comment by holke** [comment]
> Note to future self: Add link to the wiki pages to each example.

Done

File: example/tutorials/t8_step5_main.cxx
Comment (holke) [reply example/tutorials/t8_step5_main.cxx:27]: Done

File: example/tutorials/t8_step5_main.cxx
Comment (holke) [reply example/tutorials/t8_step5_main.cxx:26]: Done

File: example/tutorials/README.md
Comment (holke) [reply example/tutorials/README.md:25]: Yes, see #83

**Comment by holke** [comment]
I think i adressed all your comments now.

See also the new issues #83  and #84

**Pull request merged**
